server/tso/allocator_manager.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2020 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tso
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/election"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/embed"
+	"go.uber.org/zap"
+)
+
+type allocatorGroup struct {
+	// allocator's parent ctx and cancel function, which is to
+	// control the allocator's behavior in AllocatorDaemon and
+	// pass the cancel signal to its parent as soon as possible
+	// since it is critical to let parent goroutine know whether
+	// the allocator is still able to work well.
+	parentCtx    context.Context
+	parentCancel context.CancelFunc
+	// For the Global TSO Allocator, leadership is a PD leader's
+	// leadership, and for the Local TSO Allocator, leadership
+	// is a DC-level certificate to allow an allocator to generate
+	// TSO for local transactions in its DC.
+	leadership *election.Leadership
+	allocator  Allocator
+	// the flag indicates whether this allocator is initialized
+	isInitialized bool
+}
+
+// AllocatorManager is used to manage the TSO Allocators a PD server holds.
+// It is in charge of maintaining TSO allocators' leadership, checking election
+// priority, and forwarding TSO allocation requests to correct TSO Allocators.
+type AllocatorManager struct {
+	sync.RWMutex
+	wg sync.WaitGroup
+	// There are two kinds of TSO Allocators:
+	//   1. Global TSO Allocator, as a global single point to allocate
+	//      TSO for global transactions, such as cross-region cases.
+	//   2. Local TSO Allocator, servers for DC-level transactions.
+	// dc-location/global (string) -> TSO Allocator
+	allocatorGroups map[string]*allocatorGroup
+	// etcd and its client
+	etcd   *embed.Etcd
+	client *clientv3.Client
+	// tso config
+	rootPath      string
+	saveInterval  time.Duration
+	maxResetTSGap func() time.Duration
+}
+
+// NewAllocatorManager creates a new TSO Allocator Manager.
+func NewAllocatorManager(serverCtx context.Context, etcd *embed.Etcd, client *clientv3.Client, rootPath string, saveInterval time.Duration, maxResetTSGap func() time.Duration) *AllocatorManager {
+	allocatorManager := &AllocatorManager{
+		allocatorGroups: make(map[string]*allocatorGroup),
+		etcd:            etcd,
+		client:          client,
+		rootPath:        path.Join(rootPath, "allocator"),
+		saveInterval:    saveInterval,
+		maxResetTSGap:   maxResetTSGap,
+	}
+	return allocatorManager
+}
```

Comment:
It seems we didn't use `serverCtx context.Context` here. Do we need to handle this parameter?